File: hangman.cs
// hangman.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Threading;

class Hangman
{
    static string Colorize(string text, string color)
    {
        string col = color switch
        {
            "green" => "\x1b[92m",
            "red" => "\x1b[91m",
            "yellow" => "\x1b[93m",
            "blue" => "\x1b[94m",
            "cyan" => "\x1b[96m",
            "bold" => "\x1b[1m",
            _ => "\x1b[0m"
        };
        return col + text + "\x1b[0m";
    }

    static Dictionary<string, List<string>> WORDS = new Dictionary<string, List<string>>
    {
        {"animals", new List<string>{"кот","собака","тигр","слон","жираф","дельфин","медведь","волк","лиса","заяц"}},
        {"cities", new List<string>{"москва","париж","лондон","берлин","рим","токио","пекин","нью-йорк","каир","мехико"}},
        {"food", new List<string>{"пицца","суши","бургер","паста","салат","суп","стейк","омлет","торт","мороженое"}},
        {"professions", new List<string>{"врач","учитель","инженер","пилот","повар","художник","музыкант","архитектор","юрист","журналист"}},
        {"general", new List<string>{"программирование","алгоритм","компьютер","интернет","игра","код","данные","сервер","сайт","приложение"}}
    };

    static string[] HANGMAN_PICS = {
        @"
      +---+
          |
          |
          |
          |
          |
    =========
    ",
        @"
      +---+
      |   |
          |
          |
          |
          |
    =========
    ",
        @"
      +---+
      |   |
      O   |
          |
          |
          |
    =========
    ",
        @"
      +---+
      |   |
      O   |
      |   |
          |
          |
    =========
    ",
        @"
      +---+
      |   |
      O   |
     /|   |
          |
          |
    =========
    ",
        @"
      +---+
      |   |
      O   |
     /|\  |
          |
          |
    =========
    ",
        @"
      +---+
      |   |
      O   |
     /|\  |
     /    |
          |
    =========
    ",
        @"
      +---+
      |   |
      O   |
     /|\  |
     / \  |
          |
    =========
    "
    };

    class Stats
    {
        public int games { get; set; }
        public int wins { get; set; }
        public int losses { get; set; }
        public int streak { get; set; }
        public int best_streak { get; set; }
    }

    private int maxErrors;
    private string theme;
    private string word;
    private char[] display;
    private HashSet<char> guessed = new HashSet<char>();
    private HashSet<char> wrong = new HashSet<char>();
    private int errors;
    private bool hintUsed;
    private bool gameOver;
    private bool won;
    private Stats stats;
    private string statsFile;

    public Hangman(int maxErr, string th)
    {
        maxErrors = maxErr;
        theme = th;
        statsFile = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".hangman_stats.json");
        LoadStats();
        var words = WORDS.ContainsKey(theme) ? WORDS[theme] : WORDS["general"];
        Random rnd = new Random();
        word = words[rnd.Next(words.Count)];
        display = word.ToCharArray();
        for (int i = 0; i < display.Length; i++) display[i] = '_';
    }

    void LoadStats()
    {
        if (File.Exists(statsFile))
        {
            try
            {
                string json = File.ReadAllText(statsFile);
                stats = JsonSerializer.Deserialize<Stats>(json);
            }
            catch { stats = new Stats(); }
        }
        else stats = new Stats();
    }

    void SaveStats()
    {
        string json = JsonSerializer.Serialize(stats);
        File.WriteAllText(statsFile, json);
    }

    void DisplayState()
    {
        Console.Clear();
        Console.WriteLine(Colorize("🎯 ВИСЕЛИЦА", "bold"));
        Console.WriteLine(Colorize($"Тема: {theme}", "blue"));
        string col = errors > 0 ? "red" : "green";
        Console.WriteLine(Colorize($"Ошибок: {errors}/{maxErrors}", col));
        Console.WriteLine(HANGMAN_PICS[errors]);
        Console.Write("Слово: ");
        foreach (char c in display) Console.Write(c + " ");
        Console.WriteLine();
        if (guessed.Count > 0)
        {
            Console.Write(Colorize("Открытые буквы: ", "green"));
            Console.WriteLine(string.Join(" ", guessed));
        }
        if (wrong.Count > 0)
        {
            Console.Write(Colorize("Ошибочные: ", "red"));
            Console.WriteLine(string.Join(" ", wrong));
        }
    }

    bool Guess(char letter)
    {
        letter = char.ToLower(letter);
        if (guessed.Contains(letter) || wrong.Contains(letter))
        {
            Console.WriteLine(Colorize("Вы уже называли эту букву.", "yellow"));
            return false;
        }
        if (word.Contains(letter))
        {
            guessed.Add(letter);
            for (int i = 0; i < word.Length; i++)
                if (word[i] == letter) display[i] = letter;
            if (!display.Contains('_'))
            {
                won = true;
                gameOver = true;
            }
            return true;
        }
        else
        {
            wrong.Add(letter);
            errors++;
            if (errors >= maxErrors) gameOver = true;
            return false;
        }
    }

    void Hint()
    {
        if (hintUsed)
        {
            Console.WriteLine(Colorize("Подсказка уже использована.", "yellow"));
            return;
        }
        for (int i = 0; i < display.Length; i++)
        {
            if (display[i] == '_')
            {
                display[i] = word[i];
                guessed.Add(word[i]);
                hintUsed = true;
                Console.WriteLine(Colorize($"Подсказка: буква '{word[i]}' открыта.", "cyan"));
                if (!display.Contains('_'))
                {
                    won = true;
                    gameOver = true;
                }
                break;
            }
        }
    }

    public void Play()
    {
        Console.WriteLine(Colorize("Добро пожаловать в Виселицу!", "bold"));
        Console.WriteLine("Вводите буквы. ? - подсказка. quit - выход.");
        while (!gameOver)
        {
            DisplayState();
            Console.Write("> ");
            string cmd = Console.ReadLine().Trim();
            if (cmd == "quit") { Console.WriteLine("Выход."); return; }
            if (cmd == "?") { Hint(); continue; }
            if (cmd.Length != 1 || !char.IsLetter(cmd[0]))
            {
                Console.WriteLine(Colorize("Введите одну букву.", "red"));
                continue;
            }
            Guess(cmd[0]);
        }
        DisplayState();
        if (won)
        {
            Console.WriteLine(Colorize("🎉 Поздравляем! Вы выиграли!", "green"));
            stats.wins++;
            stats.streak++;
            if (stats.streak > stats.best_streak) stats.best_streak = stats.streak;
        }
        else
        {
            Console.WriteLine(Colorize($"💀 Вы проиграли. Загаданное слово: {word}", "red"));
            stats.losses++;
            stats.streak = 0;
        }
        stats.games++;
        SaveStats();
        Console.WriteLine(Colorize($"Статистика: игр {stats.games}, побед {stats.wins}, поражений {stats.losses}, серия {stats.streak}, лучшая {stats.best_streak}", "blue"));
    }

    static void Main(string[] args)
    {
        int maxErrors = 10;
        string theme = "general";
        bool showStats = false, resetStats = false;
        for (int i = 0; i < args.Length; i++)
        {
            string arg = args[i];
            if (arg == "easy") maxErrors = 10;
            else if (arg == "medium") maxErrors = 7;
            else if (arg == "hard") maxErrors = 5;
            else if (arg == "-t" || arg == "--theme")
            {
                if (i + 1 < args.Length)
                {
                    theme = args[++i];
                    if (!WORDS.ContainsKey(theme)) theme = "general";
                }
            }
            else if (arg == "-s" || arg == "--stats") showStats = true;
            else if (arg == "-r" || arg == "--reset") resetStats = true;
            else if (arg == "-h" || arg == "--help")
            {
                Console.WriteLine("Usage: hangman [easy|medium|hard] [-t theme] [-s] [-r]");
                return;
            }
        }
        if (resetStats)
        {
            string f = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".hangman_stats.json");
            if (File.Exists(f)) File.Delete(f);
            Console.WriteLine("Статистика сброшена.");
            return;
        }
        if (showStats)
        {
            string f = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".hangman_stats.json");
            if (File.Exists(f))
            {
                try
                {
                    string json = File.ReadAllText(f);
                    var stats = JsonSerializer.Deserialize<Stats>(json);
                    Console.WriteLine(Colorize("📊 Статистика:", "bold"));
                    Console.WriteLine($"  Игр: {stats.games}");
                    Console.WriteLine($"  Побед: {stats.wins}");
                    Console.WriteLine($"  Поражений: {stats.losses}");
                    Console.WriteLine($"  Текущая серия: {stats.streak}");
                    Console.WriteLine($"  Лучшая серия: {stats.best_streak}");
                }
                catch { Console.WriteLine("Статистика пуста."); }
            }
            else Console.WriteLine("Статистика пуста.");
            return;
        }
        Hangman game = new Hangman(maxErrors, theme);
        game.Play();
    }
}

File: hangman.java
// hangman.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.stream.*;

public class hangman {
    private static final String RESET = "\u001B[0m";
    private static final String GREEN = "\u001B[92m";
    private static final String RED = "\u001B[91m";
    private static final String YELLOW = "\u001B[93m";
    private static final String BLUE = "\u001B[94m";
    private static final String CYAN = "\u001B[96m";
    private static final String BOLD = "\u001B[1m";

    private static String colorize(String text, String color) {
        return color + text + RESET;
    }

    private static final Map<String, List<String>> WORDS = new HashMap<>();
    static {
        WORDS.put("animals", Arrays.asList("кот","собака","тигр","слон","жираф","дельфин","медведь","волк","лиса","заяц"));
        WORDS.put("cities", Arrays.asList("москва","париж","лондон","берлин","рим","токио","пекин","нью-йорк","каир","мехико"));
        WORDS.put("food", Arrays.asList("пицца","суши","бургер","паста","салат","суп","стейк","омлет","торт","мороженое"));
        WORDS.put("professions", Arrays.asList("врач","учитель","инженер","пилот","повар","художник","музыкант","архитектор","юрист","журналист"));
        WORDS.put("general", Arrays.asList("программирование","алгоритм","компьютер","интернет","игра","код","данные","сервер","сайт","приложение"));
    }

    private static final String[] HANGMAN_PICS = {
        "\n      +---+\n          |\n          |\n          |\n          |\n          |\n    =========\n",
        "\n      +---+\n      |   |\n          |\n          |\n          |\n          |\n    =========\n",
        "\n      +---+\n      |   |\n      O   |\n          |\n          |\n          |\n    =========\n",
        "\n      +---+\n      |   |\n      O   |\n      |   |\n          |\n          |\n    =========\n",
        "\n      +---+\n      |   |\n      O   |\n     /|   |\n          |\n          |\n    =========\n",
        "\n      +---+\n      |   |\n      O   |\n     /|\\  |\n          |\n          |\n    =========\n",
        "\n      +---+\n      |   |\n      O   |\n     /|\\  |\n     /    |\n          |\n    =========\n",
        "\n      +---+\n      |   |\n      O   |\n     /|\\  |\n     / \\  |\n          |\n    =========\n"
    };

    static class Stats {
        int games, wins, losses, streak, bestStreak;
    }

    private int maxErrors;
    private String theme;
    private String word;
    private char[] display;
    private Set<Character> guessed = new HashSet<>();
    private Set<Character> wrong = new HashSet<>();
    private int errors;
    private boolean hintUsed;
    private boolean gameOver;
    private boolean won;
    private Stats stats;
    private String statsFile;
    private Scanner scanner;

    public hangman(int maxErr, String th) {
        maxErrors = maxErr;
        theme = th;
        statsFile = System.getProperty("user.home") + "/.hangman_stats.json";
        loadStats();
        List<String> words = WORDS.getOrDefault(theme, WORDS.get("general"));
        Random rnd = new Random();
        word = words.get(rnd.nextInt(words.size()));
        display = word.toCharArray();
        for (int i = 0; i < display.length; i++) display[i] = '_';
        scanner = new Scanner(System.in);
    }

    private void loadStats() {
        stats = new Stats();
        try {
            String json = new String(Files.readAllBytes(Paths.get(statsFile)));
            stats.games = extractInt(json, "games");
            stats.wins = extractInt(json, "wins");
            stats.losses = extractInt(json, "losses");
            stats.streak = extractInt(json, "streak");
            stats.bestStreak = extractInt(json, "best_streak");
        } catch (Exception e) {
            // file not exists
        }
    }

    private int extractInt(String json, String key) {
        int idx = json.indexOf("\"" + key + "\"");
        if (idx == -1) return 0;
        int start = json.indexOf(":", idx) + 1;
        int end = json.indexOf(",", start);
        if (end == -1) end = json.indexOf("}", start);
        try { return Integer.parseInt(json.substring(start, end).trim()); } catch (Exception e) { return 0; }
    }

    private void saveStats() {
        try {
            String json = "{\"games\":" + stats.games + ",\"wins\":" + stats.wins +
                          ",\"losses\":" + stats.losses + ",\"streak\":" + stats.streak +
                          ",\"best_streak\":" + stats.bestStreak + "}";
            Files.write(Paths.get(statsFile), json.getBytes());
        } catch (IOException e) {}
    }

    private void displayState() {
        System.out.print("\033[H\033[2J");
        System.out.flush();
        System.out.println(colorize("🎯 ВИСЕЛИЦА", BOLD));
        System.out.println(colorize("Тема: " + theme, BLUE));
        String col = errors > 0 ? RED : GREEN;
        System.out.println(colorize("Ошибок: " + errors + "/" + maxErrors, col));
        System.out.println(HANGMAN_PICS[errors]);
        System.out.print("Слово: ");
        for (char c : display) System.out.print(c + " ");
        System.out.println();
        if (!guessed.isEmpty()) {
            System.out.print(colorize("Открытые буквы: ", GREEN));
            for (char c : guessed) System.out.print(c + " ");
            System.out.println();
        }
        if (!wrong.isEmpty()) {
            System.out.print(colorize("Ошибочные: ", RED));
            for (char c : wrong) System.out.print(c + " ");
            System.out.println();
        }
    }

    private boolean guess(char letter) {
        letter = Character.toLowerCase(letter);
        if (guessed.contains(letter) || wrong.contains(letter)) {
            System.out.println(colorize("Вы уже называли эту букву.", YELLOW));
            return false;
        }
        if (word.indexOf(letter) != -1) {
            guessed.add(letter);
            for (int i = 0; i < word.length(); i++) {
                if (word.charAt(i) == letter) display[i] = letter;
            }
            if (new String(display).indexOf('_') == -1) {
                won = true;
                gameOver = true;
            }
            return true;
        } else {
            wrong.add(letter);
            errors++;
            if (errors >= maxErrors) gameOver = true;
            return false;
        }
    }

    private void hint() {
        if (hintUsed) {
            System.out.println(colorize("Подсказка уже использована.", YELLOW));
            return;
        }
        for (int i = 0; i < display.length; i++) {
            if (display[i] == '_') {
                display[i] = word.charAt(i);
                guessed.add(word.charAt(i));
                hintUsed = true;
                System.out.println(colorize("Подсказка: буква '" + word.charAt(i) + "' открыта.", CYAN));
                if (new String(display).indexOf('_') == -1) {
                    won = true;
                    gameOver = true;
                }
                break;
            }
        }
    }

    public void play() {
        System.out.println(colorize("Добро пожаловать в Виселицу!", BOLD));
        System.out.println("Вводите буквы. ? - подсказка. quit - выход.");
        while (!gameOver) {
            displayState();
            System.out.print("> ");
            String cmd = scanner.nextLine().trim();
            if (cmd.equals("quit")) {
                System.out.println("Выход.");
                return;
            }
            if (cmd.equals("?")) {
                hint();
                continue;
            }
            if (cmd.length() != 1 || !Character.isLetter(cmd.charAt(0))) {
                System.out.println(colorize("Введите одну букву.", RED));
                continue;
            }
            guess(cmd.charAt(0));
        }
        displayState();
        if (won) {
            System.out.println(colorize("🎉 Поздравляем! Вы выиграли!", GREEN));
            stats.wins++;
            stats.streak++;
            if (stats.streak > stats.bestStreak) stats.bestStreak = stats.streak;
        } else {
            System.out.println(colorize("💀 Вы проиграли. Загаданное слово: " + word, RED));
            stats.losses++;
            stats.streak = 0;
        }
        stats.games++;
        saveStats();
        System.out.println(colorize("Статистика: игр " + stats.games + ", побед " + stats.wins +
                                    ", поражений " + stats.losses + ", серия " + stats.streak +
                                    ", лучшая " + stats.bestStreak, BLUE));
        scanner.close();
    }

    public static void main(String[] args) {
        int maxErrors = 10;
        String theme = "general";
        boolean showStats = false, resetStats = false;
        for (int i = 0; i < args.length; i++) {
            String arg = args[i];
            if (arg.equals("easy")) maxErrors = 10;
            else if (arg.equals("medium")) maxErrors = 7;
            else if (arg.equals("hard")) maxErrors = 5;
            else if (arg.equals("-t") || arg.equals("--theme")) {
                if (i+1 < args.length) {
                    theme = args[++i];
                    if (!WORDS.containsKey(theme)) theme = "general";
                }
            } else if (arg.equals("-s") || arg.equals("--stats")) showStats = true;
            else if (arg.equals("-r") || arg.equals("--reset")) resetStats = true;
            else if (arg.equals("-h") || arg.equals("--help")) {
                System.out.println("Usage: java hangman [easy|medium|hard] [-t theme] [-s] [-r]");
                return;
            }
        }
        if (resetStats) {
            String f = System.getProperty("user.home") + "/.hangman_stats.json";
            try { Files.deleteIfExists(Paths.get(f)); } catch (Exception e) {}
            System.out.println("Статистика сброшена.");
            return;
        }
        if (showStats) {
            String f = System.getProperty("user.home") + "/.hangman_stats.json";
            try {
                String json = new String(Files.readAllBytes(Paths.get(f)));
                int games = extractInt(json, "games");
                int wins = extractInt(json, "wins");
                int losses = extractInt(json, "losses");
                int streak = extractInt(json, "streak");
                int best = extractInt(json, "best_streak");
                System.out.println(colorize("📊 Статистика:", BOLD));
                System.out.println("  Игр: " + games);
                System.out.println("  Побед: " + wins);
                System.out.println("  Поражений: " + losses);
                System.out.println("  Текущая серия: " + streak);
                System.out.println("  Лучшая серия: " + best);
            } catch (Exception e) {
                System.out.println("Статистика пуста.");
            }
            return;
        }
        hangman game = new hangman(maxErrors, theme);
        game.play();
    }

    private static int extractInt(String json, String key) {
        int idx = json.indexOf("\"" + key + "\"");
        if (idx == -1) return 0;
        int start = json.indexOf(":", idx) + 1;
        int end = json.indexOf(",", start);
        if (end == -1) end = json.indexOf("}", start);
        try { return Integer.parseInt(json.substring(start, end).trim()); } catch (Exception e) { return 0; }
    }
}
